CMake fixes

SoapGentoo opened the pull request.

CelticMinstrel commented:
I'm dubious about f07f703, as those headers are included using the `<>` syntax which is supposedly what `-isystem` is supposed to be for if I recall correctly.

SoapGentoo commented:
@CelticMinstrel nope, https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html details that. `#include <bla> `is completely orthogonal to `-isystem`. See what can happen with `-isystem`:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71090
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129

loonycyborg commented:
This issue seems to apply only to using -isystem /usr/include. Using it for other directories should still be fine.

SoapGentoo commented:
Ok, `-isystem` added back

GregoryLundberg commented:
-isystem is fine and thanks for adding it back in.

@loonycyborg: The issue is not with /usr/include, per se; it is that, coupled with #include-next (as found in stdio.h and math.h), they were just the simplest examples to reproduce it.

Wesnoth's use of -isystem is consistent with the meaning and purpose of the cited documentation; we do NOT want so many diagnostic messages from upstream libraries and treating them as system headers achieves that goal. This is the entire reason GCC gives us the compiler option.

CelticMinstrel commented:
Also note that changing the name of the output file in only one toolchain isn't really a good idea. Both scons and cmake should produce a binary with the same name.

Vultraz commented:
Either we merge or close this.

SoapGentoo commented:
I'll work on it

GregoryLundberg commented:
I don't get why the bump in minimum CMake version, and why the change from configurable directory names to hard-coded names.

A number of the changes are just white-space and should be elimiated.

The change from "test" to "wesnoth-test" is probably a good idea but all it does is suppress a warning when configuring CMake. It requires a lot more than simply changing the executable name.

@Vultraz i recommend pushing this off to 1.13.10

SoapGentoo commented:
`DATAROOTDIR` is a non-standard, non-idiomatic variable that is project internal. `CMAKE_INSTALL_DATAROOTDIR` is a Kitware-supported, standardised and configurable variable for which every distribution has hooks. Using the same variables everywhere is 1) easier for everyone 2) makes maintenance easier 3) and does not require us to hack build systems.

GregoryLundberg commented:
I was referring to changes which switched from using a `{variable}` to using `{another-variable}/pathname`. `.../html`, or example, presumes `html` exists and is the desired name.

GregoryLundberg commented:
#2167 changed the target name and executable from 'test' to 'boost_unit_tests'
#2168 removed CMP005

With those both being merged to wesnoth/master CMake should run with no errors and no warnings when generating the Makefile.

I have a local subject branch for #1725 which resolves the merge conflicts and eliminates the spurious changes. I could push this up to complete work on this PR. I don't, however, see the point. It means a change to the build process for all packagers using CMake. I attempted to ask on the Forums, but garnered no responses.

My choice, at this point, is to simply close this PR, or close it in favor of my branch.

Given there seems to be no interest in the remaining changes, unless someone objects, I will be closing this PR and deleting my local branch, discarding the  remaining changes, once 1.13.11 (or 1.14 RC1) tags.

SoapGentoo commented:
@GregoryLundberg how would you like me to commit the changes, in what order?

I agree that the workflow changes for packagers, but it changes to a uniform way of doing things. Most distros already have hooks in place for `GNUInstallDirs` variables. The `GNUInstallDirs` module is the only way to let users change the dirs without each downstream reinventing the vars.

GregoryLundberg commented:
My branch has what remains of your changes, sans spurious changes; rebased as of a week or so ago. Feel free to grab it and push it over your branch; or hand-apply the changes.

If you feel the white-space changes are important, please move them to a separate commit.

I can see what you're doing and don't see a problem with it except nobody seems to care. 

So the question is merge a breaking change or not. I'm on the fence. So, if you want to stand up and state that as one of the Gentoo packagers, you want the change, I'll take that as 1 yea, 0 nay and N abstentions, and merge to see who complains. Be aware 1.13.11 (1.14 beta 3) is scheduled to tag Sunday. So if you want it, either tell me to merge my branch or update and push yours in the next day to day-and-a-half.

SoapGentoo commented:
@GregoryLundberg I've cleaned it up and rebased on master. I've also fixed the relative handling on Windows to be much simpler. Could you have another look please?

GregoryLundberg commented on CMakeLists.txt:
```diff
@@ -477,19 +474,14 @@ if(UNIX AND NOT CMAKE_COMPILER_IS_GNUCXX)
-
 if(NOT WIN32)
-	add_definitions(-DWESNOTH_PATH="${DATADIR}")
+	add_definitions(-DWESNOTH_PATH=\\\"${CMAKE_INSTALL_FULL_DATADIR}/${DATADIRNAME}\\\")
```
CMP005 has been eliminated. Remove the escapes (\\\) on both ends they are no longer required.

This is causing a Travis faliure.

GregoryLundberg commented on CMakeLists.txt:
```diff
@@ -666,15 +658,16 @@ endif()
+	set(LOCALEDIR "${CMAKE_INSTALL_FULL_LOCALEDIR}")
+endif(WIN32)
+
+add_definitions(-DLOCALEDIR=\\\"${LOCALEDIR}\\\")
```
CMP005 has been eliminated. Remove the escapes (\) on both ends they are no longer required.

This is causing a Travis faliure.

GregoryLundberg commented:
Add a separate commit to update the changelog in the (probably vain) hope the packagers read it if this breaks their workflows :P.

Rename my CI section to include packaging.
Add a comment about the CMake minimum requirements changing.
Then add a brief comment or two about what you've done.

When Travis finishes, if there are no more issues, I'll merge.

SoapGentoo commented:
@GregoryLundberg before I push, do you mind me renaming the `changelog` to `ChangeLog`? The latter is the GNU convention, which is nice, as it gets auto-installed on Gentoo and other distros.

SoapGentoo commented:
Never mind, renaming would trickle a cascade of massive packaging changes.

GregoryLundberg commented:
Travis is having problems finding the wesnoth and boost_unit_tests executable files.
I'm looking at it and think it's something your CMake changes changed which need to be reflected into utils .. I want to let Travis fully finish (which will be a couple more hours) before I dig into it. But if you see the issue (and do NOT break SCons) go ahead and push another update.

SoapGentoo commented:
@GregoryLundberg my gut feeling tells me its the removal of

    set(EXECUTABLE_OUTPUT_PATH "${CMAKE_BINARY_DIR}" [...])

The reason why I removed this is because people tend to rely too much on build tree details, which are ultimately an implementation detail of the build system and something that should not be relied upon. I would recommend installing it to some temp dir and running it from there. This also due to the use of the full data path in the executable. Your ideas?

GregoryLundberg commented:
Sounds like a strong probability. If you don't want to dirty up this PR, as a test make a [WIP, TESTING] PR (so we know not to merge it, with these changes and that change and we can compare results in a couple hours when both are through Travis.

SoapGentoo commented:
@GregoryLundberg brought previous EXECUTABLE_OUTPUT_PATH back, let's see now

GregoryLundberg commented:
Waiting for Windows VS 2013, 2015 and 2017, and macOS to finish up but there's green lights down all Linux checks. The remaining checks should all pass since they don't use CMake.

If you want to take a while and consider alternatives to the last change, fine; otherwise looks like this PR is good to go.

SoapGentoo commented:
Nah, removing `EXECUTABLE_OUTPUT_PATH` will be done in a separate PR, since it will require touching travis and a bunch of other files.

GregoryLundberg commented:
Then I will merge in a few hours when I see the CI systems pass instead of just knowing/thinking they will :)

SoapGentoo commented:
This has been in the works for 6 months (due to my timing out), I can wait another day if that is what it takes :wink:

SoapGentoo commented:
`The job exceeded the maximum time limit for jobs, and has been terminated.` :cry:

GregoryLundberg commented:
macOS timed out waiting for a Docker to free up .. restarted. np

GregoryLundberg commented:
Boom. Done.

GregoryLundberg commented:
Well crap.

You moved all the translations (*.mo) from the translations folder to the po folder.

Two problems:

1) SCons still puts them in translations
2) The PO folder already has the source files (*.po)
3) Given the above, getting .gitignore right is a pain

Ooops that's three problems.

While (1) is fixable, (2) means the packagers will need to pick apart the po directory folder/file by folder/file.

My suggestion is to put the *.po files back in the translations folder.

SoapGentoo commented:
@GregoryLundberg I'll fix it tomorrow. The reason I changed it (and the installed image looked good) is as the previous solution seemed to pollute the source tree, which is generally something that should be avoided.

GregoryLundberg commented:
Well, one developer is learning something new about fixing merge conflicts, and I learned my build farm had a glaring hole so I'm tearing it down and rebuilding the scripts. I've been meaning to do it for a couple years, anyway.

I think:

* First, a PR to revert to use translations, then

* create an Issue and PR to rename po to po-source or something like that, eliminate translations directory and build localisations in po directory. Then we can poll a few other developers into the discussion.

Vultraz commented:
I just remembered certain UMC have their own translations in `translations/'. Changing where the game looks for them could potentially mean all UMC translations break, unless we can add some fallback so the game checks in the old location too. But that seems messy and unnecessary. Suggest reverting the translations directory change.

GregoryLundberg commented:
For reference:

To be orthogonal with SCons, CMake **MUST** build the translations in the source-kit directory `~wesnoth/translations`, no matter if this is an in-tree or out-of-tree build.

Yes, all possible build workloads dirty up the source kit.

SoapGentoo commented:
@GregoryLundberg are you in the IRC channel on freenode?

Vultraz commented:
He occasionally is, under the names *tad* or *tad_carlucci*.
